lib/runtime/index.js

```diff
@@ -29,35 +32,45 @@ exports.enable = function (mod) {
   return false;
 };
 
-// Shorthand for module.watch(require(id), setters) that accepts just a
-// string module identifier `id` rather than the exports object for the
-// required module. In the future, this API will replace module.watch, and
-// will allow for creating Entry objects before module evaluation, which
-// will solve some problems with import cycles and hoisted declarations.
+// Calling module.link(id, setters) resolves the given ID using
+// module.resolve(id), which should return a canonical absolute module
+// identifier string (like require.resolve); then creates an Entry object
+// for the child module and evaluates its code (if this is the first time
+// it has been imported) by calling module.require(id). Finally, the
+// provided setter functions will be called with values exported by the
+// module, possibly multiple times when/if those exported values change.
+// The module.link name is intended to evoke the "liveness" of the
+// exported bindings, since we are subscribing to all future exports of
+// the child module, not just taking a snapshot of its current exports.
 function moduleLink(id, setters, key) {
-  return moduleWatch.call(this, this.require(id), setters, key);
-}
-
-// If key is provided, it will be used to identify the given setters so
-// that they can be replaced if module.watch is called again with the same
-// key. This avoids potential memory leaks from import declarations inside
-// loops. The compiler generates these keys automatically (and
-// deterministically) when compiling nested import declarations.
-function moduleWatch(exported, setters, key) {
   utils.setESModule(this.exports);
-  Entry.getOrCreate(this.exports, this);
+  Entry.getOrCreate(this.id, this);
+
+  var absChildId = this.resolve(id);
+  var childEntry = Entry.getOrCreate(absChildId);
 
   if (utils.isObject(setters)) {
-    Entry.getOrCreate(exported).addSetters(this, setters, key);
+    childEntry.addSetters(this, setters, key);
+  }
+
+  var exports = this.require(absChildId);
+
+  if (childEntry.module === null) {
+    childEntry.module = {
+      id: absChildId,
+      exports: exports
+    };
   }
+
+  childEntry.runSetters();
```

Comment:
Calling `childEntry.runSetters()` here is essential to report the incomplete exports of modules involved in import cycles, so that hoisted function declarations will be exposed as early as possible.